pkg/replication/replication-client_test.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	replicationlib "github.com/kube-storage/spec/lib/go/replication"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc"
+)
+
+func TestEnableVolumeReplication(t *testing.T) {
+	var client = NewReplicationClient(&grpc.ClientConn{}, time.Minute)
+	// return success response
+	mockedEnableReplication := &replicationClientMock{
+		EnableVolumeReplicationMock: func(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.EnableVolumeReplicationResponse, error) {
+			return &replicationlib.EnableVolumeReplicationResponse{}, nil
+		},
+	}
+	client = mockedEnableReplication
+	resp, err := client.EnableVolumeReplication(context.Background(), "", nil, nil)
+	assert.Equal(t, &replicationlib.EnableVolumeReplicationResponse{}, resp)
+	assert.Nil(t, err)
```

Comment:
```suggestion
	// return success response
	mockedEnableReplication := &replicationClientMock{
		EnableVolumeReplicationMock: func(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.EnableVolumeReplicationResponse, error) {
			return &replicationlib.EnableVolumeReplicationResponse{}, nil
		},
	}
	resp, err := mockedEnableReplication.EnableVolumeReplicationMock(context.Background(), "", nil, nil)
	assert.Equal(t, &replicationlib.EnableVolumeReplicationResponse{}, resp)
	assert.Nil(t, err)
```

I tried this and it works too. Is there a reason why you first do `client = NewReplicationClient()` and then overwrite it with `mockedEnableReplication` ?

Reply:
Yes, it would work but from the Reconcile logic, we cannot call mockedEnableReplication.EnableVolumeReplicationMock directly . it will call `client.EnableVolumeReplication` as we are assigning `client = mockedEnableReplication` it will ensure that when the user calls the actual GRPC client. the mock functions we are writing will get executed.

Reply:
aah makes sense. Thanks.